ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -198,28 +198,71 @@ type FSharpOverridingMembersBuilder() =
     inherit GeneratorBuilderBase<FSharpGeneratorContext>()
 
     let addNewLineIfNeeded (typeDecl: IFSharpTypeDeclaration) (typeRepr: ITypeRepresentation) =
-        if isNull typeRepr || typeDecl.StartLine <> typeRepr.StartLine then () else
+        if isNull typeRepr || typeDecl.Identifier.StartLine <> typeRepr.StartLine then () else
 
-        let currentIndent = typeRepr.Indent
-        let desiredIndent = typeDecl.Indent + typeDecl.GetIndentSize()
+        let indentSize = typeDecl.GetIndentSize()
+        let desiredIndent = typeDecl.Indent + indentSize
         
         addNodesBefore typeRepr.FirstChild [
             NewLine(typeRepr.GetLineEnding())
-            Whitespace(currentIndent)
+            Whitespace(desiredIndent)
         ] |> ignore
         
-        // to have good indents for begin/end keywords which might start out on different indents
-        shiftNode (-currentIndent) typeRepr
-        shiftNode desiredIndent typeRepr
+        let normalizeReprEnd (beginToken: ITokenNode) (endToken: ITokenNode) =
+            if beginToken.Indent > endToken.Indent then
+                let diff = beginToken.Indent - endToken.Indent
+                addNodeBefore endToken (Whitespace(diff))
         
+        match typeRepr with
+        | :? IClassRepresentation as classRepr ->
+            normalizeReprEnd classRepr.BeginKeyword classRepr.EndKeyword
+        | :? IStructRepresentation as structRepr ->
+            normalizeReprEnd structRepr.BeginKeyword structRepr.EndKeyword
+        | :? IRecordRepresentation as recordRepr ->
+            normalizeReprEnd recordRepr.LeftBrace recordRepr.RightBrace
+
+            if recordRepr.FieldDeclarations.Count > 1 then
+                recordRepr.FieldDeclarations
+                |> Seq.indexed
+                |> Seq.skip 1
+                |> Seq.iter (fun (idx, f) ->
+                    let prevField = recordRepr.FieldDeclarations[idx - 1]
+                    if f.StartLine <> prevField.StartLine && f.Indent > recordRepr.FieldDeclarations[0].Indent then
+                        let diff = f.Indent - recordRepr.FieldDeclarations[0].Indent
+                        shiftWithWhitespaceBefore -diff f)
+        | _ -> ()
+
+        // members might need deindenting after the representation start was moved to it's own line
         typeDecl.TypeMembers
```

Comment:
It seems that moving members only may produce unexpected result in cases like this one:
```fsharp
type R = { F: int }
          // comment
          member this.P = 1
```